### PR TITLE
[Track2 Function #10] Migrate functions maven plugin to app service library

### PR DESCRIPTION
**Pull request by Flanker32**
Migrate functions maven plugin to app service library

### CODE REVIEW

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java [hunk 1]
```diff
@@ -123,9 +121,38 @@ public boolean isDisableAppInsights() {
...
-        return getAzureClient().appServices().functionApps().getByResourceGroup(getResourceGroup(), getAppName());
+    private AzureAppService appServiceClient;
+
+    protected AzureAppService getOrCreateAzureAppServiceClient() throws AzureExecutionException {
+        if (appServiceClient == null) {
+            try {
+                final Account account = getAzureAccount();
+                final List<Subscription> subscriptions = account.getSubscriptions();
+                final String targetSubscriptionId = getTargetSubscriptionId(getSubscriptionId(), subscriptions, account.getSelectedSubscriptions());
+                checkSubscription(subscriptions, targetSubscriptionId);
```
Comment (shenqianjin) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AbstractFunctionMojo.java:132]: duplicated codes

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 2]
```diff
@@ -135,282 +98,298 @@
...
-                break;
-            default:
-                throw new AzureExecutionException(String.format("Unsupported runtime %s", os));
+    private Runtime getParsedRuntime() {
+        final OperatingSystem os = Optional.ofNullable(runtime.getOs()).map(OperatingSystem::fromString).orElse(OperatingSystem.WINDOWS);
+        final JavaVersion javaVersion = Optional.ofNullable(runtime.getJavaVersion()).map(JavaVersion::fromString).orElse(JavaVersion.JAVA_8);
+        return Runtime.getRuntime(os, WebContainer.JAVA_OFF, javaVersion);
+    }
+
+    private Runtime getRawRuntime() {
```
Comment (shenqianjin) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:226]: rename --> getRuntimeOrDefault

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 3]
```diff
@@ -135,282 +98,298 @@
...
+            } catch (RuntimeException e) {
+                // swallow service exception while list triggers
+            }
+            try {
+                Thread.sleep(LIST_TRIGGERS_RETRY_PERIOD_IN_SECONDS * 1000);
+            } catch (InterruptedException e) {
+                // swallow interrupted exception
+            }
         }
+        throw new AzureToolkitRuntimeException(NO_TRIGGERS_FOUNDED);
```
Comment (Flanker32) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:370]: leverage retry in utils

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 4]
```diff
@@ -135,282 +98,298 @@
...
-            final AppServicePlan appServicePlan = AppServiceUtils.getAppServicePlanByAppService(functionApp);
-            final PricingTier functionPricingTier = appServicePlan.pricingTier();
-            return PricingTier.getAll().stream().anyMatch(pricingTier -> pricingTier.equals(functionPricingTier));
-        } catch (AzureAuthFailureException e) {
-            throw new AzureExecutionException(FAILED_TO_GET_FUNCTION_APP_PRICING_TIER, e);
+    private List<FunctionEntity> listFunctions(final IFunctionApp functionApp) {
+        for (int i = 0; i < LIST_TRIGGERS_MAX_RETRY; i++) {
+            try {
+                Log.info(String.format(SYNCING_TRIGGERS_AND_FETCH_FUNCTION_INFORMATION, i + 1, LIST_TRIGGERS_MAX_RETRY));
+                functionApp.syncTriggers();
```
Comment (shenqianjin) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java:356]: retry ... logic ?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/AbstractAppService.java [hunk 5]
```diff
@@ -59,23 +59,20 @@ public void restart() {
...
     }
 
     @Override
     @Nonnull
     public R entity() {
+        if (remote == null) {
+            refresh();
+        }
         return entity;
     }
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/service/impl/AbstractAppService.java:77]: Check SDK `ManagementException`, find resource not found

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**